README.md
# limes

limes authenticates users.

## Status

| Category         | Status                                                                                                  |
| ---------------- | ------------------------------------------------------------------------------------------------------- |
| Version          | [![npm](https://img.shields.io/npm/v/limes)](https://www.npmjs.com/package/limes)                       |
| Dependencies     | ![David](https://img.shields.io/david/thenativeweb/limes)                                               |
| Dev dependencies | ![David](https://img.shields.io/david/dev/thenativeweb/limes)                                           |
| Build            | ![GitHub Actions](https://github.com/thenativeweb/limes/workflows/Release/badge.svg?branch=main) |
| License          | ![GitHub](https://img.shields.io/github/license/thenativeweb/limes)                                     |

## Installation

```shell
$ npm install limes
```

## Quick start

First you need to add a reference to limes in your application:

```javascript
const { Limes, IdentityProvider } = require('limes');
```

If you use TypeScript, use the following code instead:

```typescript
import { Limes, IdentityProvider } from 'limes';
```

Now you need to create one or more identity providers. For each identity provider call the `IdentityProvider` constructor and hand over the `issuer` as well as a `privateKey` or a `certificate`, each in `.pem` format. Optionally, you may provide both:

```javascript
const identityProvider = new IdentityProvider({
  issuer: 'https://auth.thenativeweb.io',
  privateKey: await readFile(path.join(__dirname, 'privateKey.pem')),
  certificate: await readFile(path.join(__dirname, 'certificate.pem'))
});
```

_Please note that you have to specify the private key if you want to issue tokens and the certificate if you want to verify them._

Then you can call the `Limes` constructor function to create a new limes instance. Hand over an array of the previously created identity providers:

```javascript
const limes = new Limes({
  identityProviders: [ identityProvider ]
});
```

By default, tokens issues by an identity provider are valid for 24 hours. To use a custom expiration time, provide the `expiresInMinutes` option:

```javascript
const identityProvider = new IdentityProvider({
  issuer: 'https://auth.thenativeweb.io',
  privateKey: await readFile(path.join(__dirname, 'privateKey.pem')),
  certificate: await readFile(path.join(__dirname, 'certificate.pem')),
  expiresInMinutes: 60
});
```

### Issuing tokens

To issue a token call the `issueToken` function and provide the `issuer` and the `subject` you want to use as well as an optional payload:

```javascript
const token = limes.issueToken({
  issuer: 'https://auth.thenativeweb.io',
  subject: 'jane.doe',
  payload: {
    'https://auth.thenativeweb.io/email': 'jane.doe@thenativeweb.io'
  }
});
```

_Please note that the issuer must match one of the registered identity providers. Otherwise, `issueToken` will throw an error._

#### Issuing untrusted tokens for testing

From time to time, e.g. for testing, you may want to get a JSON object that looks like a decoded token, but avoid the effort to create a signed token first. For this, use the static `issueUntrustedToken` function and hand over the desired `issuer`, the `subject`, and an optional `payload`:

```javascript
const { token, decodedToken } = Limes.issueUntrustedToken({
  issuer: 'https://untrusted.thenativeweb.io',
  subject: 'jane.doe'
});
```

_Please note that this is highly insecure, and should never be used for production code!_

### Verifying tokens

To verify a token call the `verifyToken` function and provide the token. This function tries to verify and decode the token using the identity provider that matches the token's `iss` value and returns the decoded token:

```javascript
const decodedToken = await limes.verifyToken({ token });
```

If no identity provider for the token's `iss` value is found, an exception is thrown. Also, an exception is thrown if the token is invalid.

### Using middleware

To verify tokens in web applications, there is a middleware for Express. To use it call the `verifyTokenMiddleware` function and hand over a made-up issuer value you want to use for anonymous tokens:

```javascript
app.use(limes.verifyTokenMiddleware({
  issuerForAnonymousTokens: 'https://anonymous.thenativeweb.io'
}));
```

_Please note that the issuer for anonymous tokens is made-up, and does not provide any security. It's just a string that is used without further validation._

The middleware expects the token to be inside the `authorization` HTTP header, prefixed with the term `Bearer`:

    authorization: Bearer <token>

Alternatively, you may transfer the token using the query string parameter `token`:

    GET /foo/bar?token=<token>

Either way, the verified and decoded token will be attached to the `req.user` property, while the original token will be attached to the `req.token` property:

```javascript
const app = express();

app.use(limes.verifyTokenMiddleware({
  issuerForAnonymousTokens: 'https://anonymous.thenativeweb.io'
}));

app.get('/', (req, res) => {
  res.json({ user: req.user, token: req.token });
});
```

If a request does have an invalid token, an expired one, or one from an unknown issuer, the middleware returns the status code `401`.

### Handling anonymous users

If a request does not provide a token, a token for an anonymous user will be issued. This issued token uses `anonymous` for the `sub` property, and the aforementioned issuer for anonymous tokens. Anonymous tokens have an additional claim `<issuerForAnonymousTokens>/is-anonymous` set to `true`.

_Please make sure that your application code handles anonymous users in an intended way! The middleware does not block anonymous users, it just identifies and marks them!_

To differ between multiple anonymous users, your client can send a uuid using the `X-Anonymous-Id` header:

    X-Anonymous-Id: <uuid>

Alternatively, you may pass the uuid via the query string parameter `anonymousId`:

    GET /foo/bar?anonymousId=<uuid>

This issued token uses `anonymous-<uuid>` for the `sub` property.

If both a token and an anonymous id are provided, the anonymous id is ignored.

## Running quality assurance

To run quality assurance for this module use [roboter](https://www.npmjs.com/package/roboter):

```shell
$ npx roboter
```


CHANGELOG.md
## [6.1.10](https://github.com/thenativeweb/limes/compare/6.1.9...6.1.10) (2022-02-01)


### Bug Fixes

* Update dependencies and fix type issues. ([#428](https://github.com/thenativeweb/limes/issues/428)) ([7d26816](https://github.com/thenativeweb/limes/commit/7d26816a2da86aeeb1960df642e6aad99ab9262b))

## [6.1.9](https://github.com/thenativeweb/limes/compare/6.1.8...6.1.9) (2021-08-22)


### Bug Fixes

* add include to tsconfig.json ([#404](https://github.com/thenativeweb/limes/issues/404)) ([6de7a09](https://github.com/thenativeweb/limes/commit/6de7a091cb2bd4f091ba8aa2653216cbb7d1d593))

## [6.1.8](https://github.com/thenativeweb/limes/compare/6.1.7...6.1.8) (2021-07-26)


### Bug Fixes

* Handle missing iss field. ([d73d9c6](https://github.com/thenativeweb/limes/commit/d73d9c65ae9b7d86a85b1dd15c1902fa2e74671c))

## [6.1.7](https://github.com/thenativeweb/limes/compare/6.1.6...6.1.7) (2021-06-10)


### Bug Fixes

* bump normalize-url from 6.0.0 to 6.0.1 ([#379](https://github.com/thenativeweb/limes/issues/379)) ([1bb8b3c](https://github.com/thenativeweb/limes/commit/1bb8b3cbd5f08530a09d2b8581c9ee17df3ae3c0))

## [6.1.6](https://github.com/thenativeweb/limes/compare/6.1.5...6.1.6) (2021-06-08)


### Bug Fixes

* bump trim-newlines from 3.0.0 to 3.0.1 ([#377](https://github.com/thenativeweb/limes/issues/377)) ([141a8d2](https://github.com/thenativeweb/limes/commit/141a8d240f6e8968cd054da60b94263eb29a9e00))

## [6.1.5](https://github.com/thenativeweb/limes/compare/6.1.4...6.1.5) (2021-06-08)


### Bug Fixes

* bump glob-parent from 5.1.1 to 5.1.2 ([#378](https://github.com/thenativeweb/limes/issues/378)) ([8603b55](https://github.com/thenativeweb/limes/commit/8603b551a73bdcb7b33c44508ee6277163d293d0))

## [6.1.4](https://github.com/thenativeweb/limes/compare/6.1.3...6.1.4) (2021-05-31)


### Bug Fixes

* bump @types/express from 4.17.11 to 4.17.12 ([#370](https://github.com/thenativeweb/limes/issues/370)) ([991cd31](https://github.com/thenativeweb/limes/commit/991cd3181f9b2000b13d56dd6800e6ff6ce6776a))

## [6.1.3](https://github.com/thenativeweb/limes/compare/6.1.2...6.1.3) (2021-05-10)


### Bug Fixes

* bump hosted-git-info from 2.8.8 to 2.8.9 ([#361](https://github.com/thenativeweb/limes/issues/361)) ([ce4fdfd](https://github.com/thenativeweb/limes/commit/ce4fdfdc3caebe4c734ad235025cb1a908056d49))

## [6.1.2](https://github.com/thenativeweb/limes/compare/6.1.1...6.1.2) (2021-03-25)


### Bug Fixes

* Migrate from master to main. ([#330](https://github.com/thenativeweb/limes/issues/330)) ([a6ecf0b](https://github.com/thenativeweb/limes/commit/a6ecf0b03dfeb279c49bfe50e1ca611875436773))

## [6.1.1](https://github.com/thenativeweb/limes/compare/6.1.0...6.1.1) (2021-03-19)


### Bug Fixes

* bump @types/jsonwebtoken from 8.5.0 to 8.5.1 ([#321](https://github.com/thenativeweb/limes/issues/321)) ([e1cdb04](https://github.com/thenativeweb/limes/commit/e1cdb0427876d057992ddcff55c556e0fda0582c))

# [6.1.0](https://github.com/thenativeweb/limes/compare/6.0.9...6.1.0) (2021-02-10)


### Features

* Add logger ([#277](https://github.com/thenativeweb/limes/issues/277)) ([f684300](https://github.com/thenativeweb/limes/commit/f684300b9c2a4a55f9b2ee3f59e1f2e302c46816))

## [6.0.9](https://github.com/thenativeweb/limes/compare/6.0.8...6.0.9) (2021-01-13)


### Bug Fixes

* bump @types/express from 4.17.10 to 4.17.11 ([#276](https://github.com/thenativeweb/limes/issues/276)) ([ad776d9](https://github.com/thenativeweb/limes/commit/ad776d95c1670925c774e13efce81c644efe2f65))

## [6.0.8](https://github.com/thenativeweb/limes/compare/6.0.7...6.0.8) (2021-01-11)


### Bug Fixes

* bump @types/express from 4.17.9 to 4.17.10 ([#274](https://github.com/thenativeweb/limes/issues/274)) ([af5b181](https://github.com/thenativeweb/limes/commit/af5b18126d2fce23931433ce07b99e8f3f6c800c))

## [6.0.7](https://github.com/thenativeweb/limes/compare/6.0.6...6.0.7) (2020-11-11)


### Bug Fixes

* bump @types/express from 4.17.8 to 4.17.9 ([#241](https://github.com/thenativeweb/limes/issues/241)) ([7f85445](https://github.com/thenativeweb/limes/commit/7f8544564000bb3e3aef4806a25e2ec8a9947253))

## [6.0.6](https://github.com/thenativeweb/limes/compare/6.0.5...6.0.6) (2020-11-03)


### Bug Fixes

* Fix headline for robot section in readme. ([#225](https://github.com/thenativeweb/limes/issues/225)) ([9447fae](https://github.com/thenativeweb/limes/commit/9447fae62e80bca0a34419e499d5cc6e057cdaf3))

## [6.0.5](https://github.com/thenativeweb/limes/compare/6.0.4...6.0.5) (2020-09-02)


### Bug Fixes

* bump @types/express from 4.17.7 to 4.17.8 ([#195](https://github.com/thenativeweb/limes/issues/195)) ([67f8531](https://github.com/thenativeweb/limes/commit/67f85317830707b0c655adcb1f7f4e4a9ded5d63))

## [6.0.4](https://github.com/thenativeweb/limes/compare/6.0.3...6.0.4) (2020-07-08)


### Bug Fixes

* bump @types/express from 4.17.6 to 4.17.7 ([#169](https://github.com/thenativeweb/limes/issues/169)) ([a7c8f31](https://github.com/thenativeweb/limes/commit/a7c8f31913b1d6934dfe4ca0a561c16a237e0939))

## [6.0.3](https://github.com/thenativeweb/limes/compare/6.0.2...6.0.3) (2020-05-31)


### Bug Fixes

* Fix dependencies. ([65cde1e](https://github.com/thenativeweb/limes/commit/65cde1ee271df2fc4dbb05704d8a9daebcf1d038))

## [6.0.2](https://github.com/thenativeweb/limes/compare/6.0.1...6.0.2) (2020-02-18)


### Bug Fixes

* Remove the need for an identity provider. ([#111](https://github.com/thenativeweb/limes/issues/111)) ([df6eb4b](https://github.com/thenativeweb/limes/commit/df6eb4bf87fd54102087ae2347c48e96dde2e54c))


test/LimesTests.js
'use strict';

var fs = require('fs'),
    http = require('http'),
    path = require('path');

var assert = require('assertthat'),
    express = require('express'),
    jwt = require('jsonwebtoken'),
    request = require('supertest'),
    socketIo = require('socket.io'),
    socketIoClient = require('socket.io-client');

var Limes = require('../lib/Limes');

var certificate = fs.readFileSync(path.join(__dirname, 'keys', 'certificate.pem')),
    privateKey = fs.readFileSync(path.join(__dirname, 'keys', 'privateKey.pem'));

suite('Limes', function () {
  test('is a function.', function (done) {
    assert.that(Limes).is.ofType('function');
    done();
  });

  test('throws an exception if options are missing.', function (done) {
    assert.that(function () {
      /* eslint-disable no-new */
      new Limes();
      /* eslint-enable no-new */
    }).is.throwing('Options are missing.');
    done();
  });

  test('throws an exception if identity provider name is missing.', function (done) {
    assert.that(function () {
      /* eslint-disable no-new */
      new Limes({});
      /* eslint-enable no-new */
    }).is.throwing('Identity provider name is missing.');
    done();
  });

  test('throws an exception if private key and certificate are both missing.', function (done) {
    assert.that(function () {
      /* eslint-disable no-new */
      new Limes({
        identityProviderName: 'auth.example.com'
      });
      /* eslint-enable no-new */
    }).is.throwing('Specify private key and / or certificate.');
    done();
  });

  suite('issueTokenFor', function () {
    test('is a function.', function (done) {
      var limes = new Limes({
        identityProviderName: 'auth.example.com',
        privateKey: privateKey,
        certificate: certificate
      });

      assert.that(limes.issueTokenFor).is.ofType('function');
      done();
    });

    test('throws an exception if subject is missing.', function (done) {
      var limes = new Limes({
        identityProviderName: 'auth.example.com',
        privateKey: privateKey,
        certificate: certificate
      });

      assert.that(function () {
        limes.issueTokenFor();
      }).is.throwing('Subject is missing.');
      done();
    });

    test('returns a JWT.', function (done) {
      var limes = new Limes({
        identityProviderName: 'auth.example.com',
        privateKey: privateKey,
        certificate: certificate
      });

      var token = limes.issueTokenFor('test.domain.com', { foo: 'bar' });

      jwt.verify(token, certificate, { issuer: 'auth.example.com' }, function (err, decodedToken) {
        assert.that(err).is.null();
        assert.that(decodedToken.iss).is.equalTo('auth.example.com');
        assert.that(decodedToken.sub).is.equalTo('test.domain.com');
        assert.that(decodedToken.foo).is.equalTo('bar');
        done();
      });
    });
  });

  suite('issueTokenForAnonymous', function () {
    test('is a function.', function (done) {
      var limes = new Limes({
        identityProviderName: 'auth.example.com',
        privateKey: privateKey,
        certificate: certificate
      });

      assert.that(limes.issueTokenForAnonymous).is.ofType('function');
      done();
    });

    test('returns a JWT.', function (done) {
      var limes = new Limes({
        identityProviderName: 'auth.example.com',
        privateKey: privateKey,
        certificate: certificate
      });

      var token = limes.issueTokenForAnonymous({ foo: 'bar' });

      jwt.verify(token, certificate, { issuer: 'auth.example.com' }, function (err, decodedToken) {
        assert.that(err).is.null();
        assert.that(decodedToken.iss).is.equalTo('auth.example.com');
        assert.that(decodedToken.sub).is.equalTo('anonymous');
        assert.that(decodedToken.foo).is.equalTo('bar');
        done();
      });
    });
  });

  suite('verifyToken', function () {
    test('is a function.', function (done) {
      var limes = new Limes({
        identityProviderName: 'auth.example.com',
        privateKey: privateKey,
        certificate: certificate
      });

      assert.that(limes.verifyToken).is.ofType('function');
      done();
    });

    test('returns the decoded token if the token is valid.', function (done) {
      var limes = new Limes({
        identityProviderName: 'auth.example.com',
        privateKey: privateKey,
        certificate: certificate
      });

      var token = limes.issueTokenFor('adc225b7-65b9-48f4-be4d-c5108aa4d1f4', {
        foo: 'bar'
      });

      limes.verifyToken(token, function (err, decodedToken) {
        assert.that(err).is.null();
        assert.that(decodedToken.iss).is.equalTo('auth.example.com');
        assert.that(decodedToken.sub).is.equalTo('adc225b7-65b9-48f4-be4d-c5108aa4d1f4');
        assert.that(decodedToken.foo).is.equalTo('bar');
        done();
      });
    });

    test('returns an error if the token is not valid.', function (done) {
      var limes = new Limes({
        identityProviderName: 'auth.example.com',
        privateKey: privateKey,
        certificate: certificate
      });

      limes.verifyToken('invalidtoken', function (err) {
        assert.that(err).is.not.null();
        done();
      });
    });

    test('returns an error if the token contains invalid characters.', function (done) {
      var limes = new Limes({
        identityProviderName: 'auth.example.com',
        privateKey: privateKey,
        certificate: certificate
      });

      limes.verifyToken('invalid token', function (err) {
        assert.that(err).is.not.null();
        done();
      });
    });
  });

  suite('middleware integration', function () {
    var limesInThePast,
        limesOfSomebodyElse;

    suiteSetup(function () {
      limesInThePast = new Limes({
        identityProviderName: 'auth.example.com',
        privateKey: privateKey,
        certificate: certificate,
        expiresInMinutes: -5
      });

      limesOfSomebodyElse = new Limes({
        identityProviderName: 'somebodyelse.example.com',
        privateKey: privateKey,
        certificate: certificate
      });
    });

    suite('verifyTokenMiddlewareExpress', function () {
      test('is a function.', function (done) {
        var limes = new Limes({
          identityProviderName: 'auth.example.com',
          privateKey: privateKey,
          certificate: certificate
        });

        assert.that(limes.verifyTokenMiddlewareExpress).is.ofType('function');
        done();
      });

      test('returns a function.', function (done) {
        var limes = new Limes({
          identityProviderName: 'auth.example.com',
          privateKey: privateKey,
          certificate: certificate
        });

        var middleware = limes.verifyTokenMiddlewareExpress();

        assert.that(middleware).is.ofType('function');
        done();
      });

      suite('middleware', function () {
        var app,
            limes;

        setup(function () {
          limes = new Limes({
            identityProviderName: 'auth.example.com',
            privateKey: privateKey,
            certificate: certificate
          });

          app = express();
          app.use(limes.verifyTokenMiddlewareExpress({
            payloadWhenAnonymous: {
              foo: 'anonymous-bar'
            }
          }));
          app.get('/', function (req, res) {
            res.send(req.user);
          });
        });

        test('returns an anonymous token for non-authenticated requests.', function (done) {
          request(app).
            get('/').
            set('accept', 'application/json').
            end(function (err, res) {
              assert.that(err).is.null();
              assert.that(res.statusCode).is.equalTo(200);
              assert.that(res.body.iss).is.equalTo('auth.example.com');
              assert.that(res.body.sub).is.equalTo('anonymous');
              assert.that(res.body.foo).is.equalTo('anonymous-bar');
              done();
            });
        });

        test('returns 401 for invalid authenticated requests.', function (done) {
          request(app).
            get('/').
            set('accept', 'application/json').
            set('authorization', 'Bearer invalidtoken').
            end(function (err, res) {
              assert.that(err).is.null();
              assert.that(res.statusCode).is.equalTo(401);
              done();
            });
        });

        test('returns 401 for tokens with invalid characters.', function (done) {
          request(app).
            get('/').
            set('accept', 'application/json').
            set('authorization', 'Bearer invalid token').
            end(function (err, res) {
              assert.that(err).is.null();
              assert.that(res.statusCode).is.equalTo(401);
              done();
            });
        });

        test('returns 401 for expired requests.', function (done) {
          var expiredToken = limesInThePast.issueTokenFor('test.domain.com', {
            foo: 'authenticated-bar'
          });

          request(app).
            get('/').
            set('accept', 'application/json').
            set('authorization', 'Bearer ' + expiredToken).
            end(function (err, res) {
              assert.that(err).is.null();
              assert.that(res.statusCode).is.equalTo(401);
              done();
            });
        });

        test('returns 401 for authenticated requests with wrong issuer.', function (done) {
          var token = limesOfSomebodyElse.issueTokenFor('test.domain.com', {
            foo: 'authenticated-bar'
          });

          request(app).
            get('/').
            set('accept', 'application/json').
            set('authorization', 'Bearer ' + token).
            end(function (err, res) {
              assert.that(err).is.null();
              assert.that(res.statusCode).is.equalTo(401);
              done();
            });
        });

        test('returns a decoded token for authenticated requests.', function (done) {
          var token = limes.issueTokenFor('test.domain.com', {
            foo: 'authenticated-bar'
          });

          request(app).
            get('/').
            set('accept', 'application/json').
            set('authorization', 'Bearer ' + token).
            end(function (err, res) {
              assert.that(err).is.null();
              assert.that(res.statusCode).is.equalTo(200);
              assert.that(res.body.iss).is.equalTo('auth.example.com');
              assert.that(res.body.sub).is.equalTo('test.domain.com');
              assert.that(res.body.foo).is.equalTo('authenticated-bar');
              done();
            });
        });

        test('returns a decoded token for authenticated requests using query string.', function (done) {
          var token = limes.issueTokenFor('test.domain.com', {
            foo: 'authenticated-bar'
          });

          request(app).
            get('/?token=' + token).
            set('accept', 'application/json').
            end(function (err, res) {
              assert.that(err).is.null();
              assert.that(res.statusCode).is.equalTo(200);
              assert.that(res.body.iss).is.equalTo('auth.example.com');
              assert.that(res.body.sub).is.equalTo('test.domain.com');
              assert.that(res.body.foo).is.equalTo('authenticated-bar');
              done();
            });
        });
      });
    });

    suite('verifyTokenMiddlewareSocketIo', function () {
      test('is a function.', function (done) {
        var limes = new Limes({
          identityProviderName: 'auth.example.com',
          privateKey: privateKey,
          certificate: certificate
        });

        assert.that(limes.verifyTokenMiddlewareSocketIo).is.ofType('function');
        done();
      });

      test('returns a function.', function (done) {
        var limes = new Limes({
          identityProviderName: 'auth.example.com',
          privateKey: privateKey,
          certificate: certificate
        });

        var middleware = limes.verifyTokenMiddlewareSocketIo();

        assert.that(middleware).is.ofType('function');
        done();
      });

      suite('middleware', function () {
        var limes;

        suiteSetup(function () {
          var app,
              io,
              server;

          limes = new Limes({
            identityProviderName: 'auth.example.com',
            privateKey: privateKey,
            certificate: certificate
          });

          app = express();
          server = http.createServer(app);
          io = socketIo.listen(server);
          io.use(limes.verifyTokenMiddlewareSocketIo({
            payloadWhenAnonymous: {
              foo: 'anonymous-bar'
            }
          }));
          io.on('connection', function (socket) {
            socket.on('getUser', function (callback) {
              callback(socket.user);
            });
          });
          server.listen(3000);
        });

        test('returns an anonymous token for non-authenticated requests.', function (done) {
          var socket = socketIoClient.connect('http://localhost:3000', { forceNew: true });

          socket.once('connect', function () {
            socket.emit('getUser', function (token) {
              assert.that(token.iss).is.equalTo('auth.example.com');
              assert.that(token.sub).is.equalTo('anonymous');
              assert.that(token.foo).is.equalTo('anonymous-bar');
              socket.disconnect();
              done();
            });
          });
        });

        test('returns an error for invalid authenticated requests.', function (done) {
          var socket = socketIoClient.connect('http://localhost:3000', { forceNew: true });

          socket.once('connect', function () {
            socket.emit('authenticate', 'invalidtoken', function (err) {
              assert.that(err).is.not.null();
              socket.disconnect();
              done();
            });
          });
        });

        test('returns an error for expired requests.', function (done) {
          var expiredToken = limesInThePast.issueTokenFor('test.domain.com', {
            foo: 'authenticated-bar'
          });

          var socket = socketIoClient.connect('http://localhost:3000', { forceNew: true });

          socket.once('connect', function () {
            socket.emit('authenticate', expiredToken, function (err) {
              assert.that(err).is.not.null();
              socket.disconnect();
              done();
            });
          });
        });

        test('returns an error for authenticated requests with wrong issuer.', function (done) {
          var token = limesOfSomebodyElse.issueTokenFor('test.domain.com', {
            foo: 'authenticated-bar'
          });

          var socket = socketIoClient.connect('http://localhost:3000', { forceNew: true });

          socket.once('connect', function () {
            socket.emit('authenticate', token, function (err) {
              assert.that(err).is.not.null();
              socket.disconnect();
              done();
            });
          });
        });

        test('returns a decoded token for authenticated requests.', function (done) {
          var token = limes.issueTokenFor('test.domain.com', {
            foo: 'authenticated-bar'
          });

          var socket = socketIoClient.connect('http://localhost:3000', { forceNew: true });

          socket.once('connect', function () {
            socket.emit('authenticate', token, function (err) {
              assert.that(err).is.null();
              socket.emit('getUser', function (receivedToken) {
                assert.that(receivedToken.iss).is.equalTo('auth.example.com');
                assert.that(receivedToken.sub).is.equalTo('test.domain.com');
                assert.that(receivedToken.foo).is.equalTo('authenticated-bar');
                socket.disconnect();
                done();
              });
            });
          });
        });
      });
    });
  });
});


lib/Limes.js
'use strict';

const jwt = require('jsonwebtoken');

const IdentityProvider = require('./IdentityProvider');

class Limes {
  constructor ({ identityProviders }) {
    if (!identityProviders) {
      throw new Error('Identity providers are missing.');
    }
    if (identityProviders.length === 0) {
      throw new Error('Identity providers are missing.');
    }

    this.identityProviders = identityProviders;
  }

  getIdentityProviderByIssuer ({ issuer }) {
    if (!issuer) {
      throw new Error('Issuer is missing.');
    }

    const requestedIdentityProvider = this.identityProviders.find(
      identityProvider => identityProvider.issuer === issuer
    );

    if (!requestedIdentityProvider) {
      throw new Error(`Issuer '${issuer}' not found.`);
    }

    return requestedIdentityProvider;
  }

  issueToken ({ issuer, subject, payload = {}}) {
    if (!issuer) {
      throw new Error('Issuer is missing.');
    }
    if (!subject) {
      throw new Error('Subject is missing.');
    }

    const identityProvider = this.getIdentityProviderByIssuer({ issuer });

    const token = jwt.sign(payload, identityProvider.privateKey, {
      algorithm: 'RS256',
      subject,
      issuer: identityProvider.issuer,
      expiresIn: identityProvider.expiresInMinutes * 60
    });

    return token;
  }

  static issueUntrustedToken ({ issuer, subject, payload = {}}) {
    if (!issuer) {
      throw new Error('Issuer is missing.');
    }
    if (!subject) {
      throw new Error('Subject is missing.');
    }

    const expiresInMinutes = 60;

    const token = jwt.sign(payload, null, {
      algorithm: 'none',
      subject,
      issuer,
      expiresIn: expiresInMinutes * 60
    });

    const decodedToken = jwt.decode(token);

    return { token, decodedToken };
  }

  async verifyToken ({ token }) {
    if (!token) {
      throw new Error('Token is missing.');
    }

    let untrustedDecodedToken;

    try {
      untrustedDecodedToken = jwt.decode(token);
    } catch {
      throw new Error('Failed to verify token.');
    }

    if (!untrustedDecodedToken) {
      throw new Error('Failed to verify token.');
    }

    const identityProvider = this.getIdentityProviderByIssuer({
      issuer: untrustedDecodedToken.iss
    });

    const decodedToken = await new Promise((resolve, reject) => {
      try {
        jwt.verify(token, identityProvider.certificate, {
          algorithms: [ 'RS256' ],
          issuer: identityProvider.issuer
        }, (err, verifiedToken) => {
          if (err) {
            return reject(new Error('Failed to verify token.'));
          }

          resolve(verifiedToken);
        });
      } catch (ex) {
        reject(ex);
      }
    });

    return decodedToken;
  }

  verifyTokenMiddleware ({ issuerForAnonymousTokens }) {
    if (!issuerForAnonymousTokens) {
      throw new Error('Issuer for anonymous tokens is missing.');
    }

    return async (req, res, next) => {
      let token;

      const authorizationHeader = req.headers.authorization,
            authorizationQuery = req.query.token;

      if (authorizationHeader) {
        const [ authorizationType, authorizationValue ] = authorizationHeader.split(' ');

        if (authorizationType === 'Bearer') {
          token = authorizationValue;
        }
      } else if (authorizationQuery) {
        token = authorizationQuery;
      }

      let decodedToken;

      if (token) {
        try {
          decodedToken = await this.verifyToken({ token });
        } catch {
          return res.status(401).end();
        }
      } else {
        ({ token, decodedToken } = Limes.issueUntrustedToken({
          issuer: issuerForAnonymousTokens,
          subject: 'anonymous'
        }));
      }

      /* eslint-disable no-param-reassign */
      req.token = token;
      req.user = {
        id: decodedToken.sub,
        claims: decodedToken
      };
      /* eslint-enable no-param-reassign */

      next();
    };
  }
}

Limes.IdentityProvider = IdentityProvider;

module.exports = Limes;
